ci.py

```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/env python3
+from pathlib import Path
+
+import argparse
+import configparser
+import logging
+import os
+import sys
+import pprint
+from subprocess import check_output, check_call, call, CalledProcessError
+
+import yaml
+import git
+
+logging.basicConfig(stream=sys.stdout, format='[%(levelname)s] %(asctime)s: %(message)s', level=logging.INFO)
+
+DEFAULT_BRANCH = 'main' # we can rewrite it in config
+YAML_CONFIG = 'ci-default.yml'
+YAML_CONFIG_CUSTOM = 'ci.yml'
+SUBMODULES_CONFIG = '.gitmodules'
+GIT_SOURCES_FILE = '.git-sources'
+
+
+class Builder():
+    rootdir = check_output(["git", "rev-parse", "--show-toplevel"]).decode('utf-8').strip()
+
+    def __init__(self):
+        self.global_branch_name = None
+        self.config = {}
+        self.custom_config = {}
+        self.read_custom_config()
+        self.read_config()
+        self.deps = self.config['deps']
+
+    def read_custom_config(self):
+        with open(YAML_CONFIG_CUSTOM, 'r') as f:
+            self.custom_config = yaml.load(f, Loader=yaml.FullLoader)
+
+    def write_custom_config(self, config):
+        with open(YAML_CONFIG_CUSTOM, 'w') as f:
+            yaml.dump(config, f, sort_keys=False)
+
+
+    def read_config(self):
+        with open(YAML_CONFIG, 'r') as f:
+            self.config = yaml.load(f, Loader=yaml.FullLoader)
+
+        if self.custom_config is not None:
+            # first we want to find global branch
+            for conf in self.custom_config['deps']:
+                if conf['name'] == 'global':
+                    self.global_branch_name = conf['branch']
+                    self.set_global_branches()
+                    break
+
+            # Yep we have high complexity here but list is short
+            for conf in self.custom_config['deps']:
+                if conf['name'] == 'global':
+                    continue
+                for dep in self.config['deps']:
+                    if dep['name'] == conf['name']:
+                        # TODO add support for other fields
+                        dep['branch'] = conf['branch']
+                        break
+                else:
+                    logging.error(f'Can"t find {conf["name"]} repo. We have list of repos in ci-default.yml')
+                    sys.exit(1)
+
+    def set_global_branches(self):
+        for dep in self.config['deps']:
+            url = dep['url']
+            g = git.cmd.Git()
+            # TODO maybe it'll be faster to use local data
+            output = g.ls_remote("--heads", url, self.global_branch_name)
+            if self.global_branch_name in output:
+                logging.info(f'Use branch {self.global_branch_name} for {dep["name"]}')
+                dep['branch'] = self.global_branch_name
+
+    def create_fb(self, branch_name):
```

Comment:
`create_fb` doesn't create FB actually, right? Maybe it's better to rename it to `prepare_deps`?

Reply:
It creates a branch for FB. We only need to create PR but I hope we'll implement it too.

Reply:
`create_fb_branch` then. Code should be self-explaining, do not mislead the ones who will read the code.
Creating branch for FB is a part of actual creating FB process which includes:
1. Create FB branch
2. Create PR
3. Build FB

Reply:
Renamed.